bridge/ffi/src/c_api.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{bindings::*, utils::*};
+use dsnp_graph_config::SchemaId;
+use dsnp_graph_core::{
+	api::api::{GraphAPI, GraphState},
+	dsnp::dsnp_types::DsnpUserId,
+};
+use std::{mem::ManuallyDrop, panic, sync::Mutex};
+
+#[no_mangle]
+pub extern "C" fn print_hello_graph() {
+	println!("Hello, Graph!");
+}
+
+// Collection of GraphStates
+#[no_mangle]
+static GRAPH_STATES: Mutex<Vec<Box<GraphState>>> = Mutex::new(Vec::new());
+
+// Initialize GraphState
+#[no_mangle]
+pub unsafe extern "C" fn initialize_graph_state(
+	environment: *const Environment,
+) -> *mut GraphState {
+	// use panic_unwind
+	let result = panic::catch_unwind(|| {
+		let environment = &*environment;
+		let rust_environment = environment_from_ffi(environment);
+		let graph_state = Box::new(GraphState::new(rust_environment));
+		let graph_state_ptr = Box::into_raw(graph_state);
+
+		// Add state pointer to GRAPH_STATES vector
+		GRAPH_STATES.lock().unwrap().push(unsafe { Box::from_raw(graph_state_ptr) });
+
+		graph_state_ptr
+	});
+	match result {
+		Ok(graph_state) => graph_state,
+		Err(_) => std::ptr::null_mut(),
+	}
+}
+
+// Intialize GraphState with capacity
+#[no_mangle]
+pub unsafe extern "C" fn initialize_graph_state_with_capacity(
+	environment: *const Environment,
+	capacity: usize,
+) -> *mut GraphState {
+	let result = panic::catch_unwind(|| {
+		let environment = &*environment;
+		let rust_environment = environment_from_ffi(environment);
+		let graph_state = Box::new(GraphState::with_capacity(rust_environment, capacity));
+		let graph_state_ptr = Box::into_raw(graph_state);
+
+		// Add state pointer to GRAPH_STATES vector
+		GRAPH_STATES.lock().unwrap().push(unsafe { Box::from_raw(graph_state_ptr) });
+
+		graph_state_ptr
+	});
+	match result {
+		Ok(graph_state) => graph_state,
+		Err(_) => std::ptr::null_mut(),
+	}
+}
+
+// Get Capacity
+#[no_mangle]
+pub unsafe extern "C" fn get_graph_capacity(graph_state: *mut GraphState) -> usize {
+	if graph_state.is_null() {
+		return 0
+	}
+	let graph_state = &mut *graph_state;
+	graph_state.capacity()
+}
+
+// Get total graph states in GRAPH_STATES
+#[no_mangle]
+pub unsafe extern "C" fn get_graph_states_count() -> usize {
+	let graph_states = GRAPH_STATES.lock().unwrap();
+	graph_states.len()
+}
+
+// State contains user graph
+#[no_mangle]
+pub unsafe extern "C" fn graph_contains_user(
+	graph_state: *mut GraphState,
+	user_id: *const DsnpUserId,
+) -> bool {
+	if graph_state.is_null() {
+		return false
+	}
+	let graph_state = &mut *graph_state;
+	let user_id = &*user_id;
+	graph_state.contains_user_graph(user_id)
```

Comment:
so I think the `catch_unwind` should be wrapping all api calls since you would not know which one might panic. You can implement something like the following to pass in a lambda function on all api calls
https://github.com/signalapp/libsignal/blob/0e2f87f39a2ebf8cc6e30fbff7fbf2c207b820da/rust/bridge/shared/src/ffi/mod.rs#L82-L95

Reply:
Now that GraphState is thread safe, panic_unwind wont complain

https://github.com/LibertyDSNP/graph-sdk/pull/57/commits/27d24a2b6c5b0598a73ee017d6e3d3e6eb4e60f2